spokestack/asr/cloud_client.py

```diff
@@ -0,0 +1,206 @@
+            self._key, self._body.encode("utf-8"), hashlib.sha256
+        ).digest()
+        self._signature = base64.b64encode(signature).decode("utf-8")
+        self._socket: Any = None
+
+        self._response: Dict[str, Any] = {
+            "error": None,
+            "final": False,
+            "hypotheses": [],
+            "status": None,
+        }
+        self._sample_rate: int = sample_rate
+        self._is_final: bool = False
+        self._idle_timeout = idle_timeout
+        self._idle_count: int = 0
+
+    def __call__(self, audio: np.ndarray, n_best: int = 1) -> List[str]:
+        """ Audio to text interface for the cloud client
+
+        Args:
+            audio (np.ndarray): np.int16 array of audio
+            n_best (int): number of predictions to return
+
+        Returns: n_best list of transcripts
+
+        """
+
+        chunk_size = 2 * int(0.01 * self._sample_rate)
+        self.connect()
+        self.initialize()
+
+        for i in range(0, len(audio), chunk_size):
+            frame = audio[i:][:chunk_size]
+            self.send(frame)
+            self.receive()
+
+        self.end()
+        while not self._is_final:
+            if self._idle_timeout and (self._idle_count > self._idle_timeout):
+                break
+            else:
+                self.receive()
+                self._idle_count += 1
+        self.disconnect()
+
+        transcripts = []
+        hypotheses = self._response.get("hypotheses")
+        if hypotheses:
+            for hypothesis in hypotheses[:n_best]:
+                transcripts.append(hypothesis.get("transcript"))
+
+        return transcripts
+
+    @property
+    def is_connected(self) -> bool:
+        """ status of the socket connection """
+        if self._socket:
+            return True
+        return False
+
+    def close(self) -> None:
+        """ closes socket """
+        self._socket.close()
+
+    def connect(self) -> None:
+        """ connects to websocket """
+        if self._socket:
+            pass
+        else:
+            self._socket = WebSocket()
+            self._socket.connect(f"{self._socket_url}/v1/asr/websocket")
+
+    def initialize(self) -> None:
+        """ sends/receives the initial api request """
+        if not self._socket:
+            raise ConnectionError("Not Connected")
+
+        message = {
+            "keyId": self._key_id,
+            "signature": self._signature,
+            "body": self._body,
+        }
+        self._socket.send(json.dumps(message))
+        self._response = json.loads(self._socket.recv())
+        self._is_final = False
+        if not self._response["status"] == "ok":
+            raise APIError(self._response)
+
+    def disconnect(self) -> None:
+        """ disconnects client socket connection """
+        if self._socket:
+            self.close()
+        self._socket = None
+
+    def send(self, frame):
+        """ sends a single frame of binary audio """
```

Comment:
`frame` should have a mypy type, and the docstring should indicate this
also, it looks like the value that is being passed in here is a `np.array`, so the docstring shouldn't indicate that the value is a binary
and I'm guessing that the websocket stuff is transparently converting the array to bytes, but it would be better to make this explicit by calling `to_bytes` here